src/main/java/org/apache/sysml/hops/rewrite/RewriteEMult.java

```diff
@@ -0,0 +1,284 @@
+			sorted.put(h, entry.getCount());
+		}
+		// sorted contains all leaves, sorted by data type, stripped from their parents
+
+		// Construct right-deep EMult tree
+		final Iterator<Map.Entry<Hop, Integer>> iterator = sorted.entrySet().iterator();
+		Hop first = constructPower(iterator.next());
+
+		for (int i = 1; i < sorted.size(); i++) {
+			final Hop second = constructPower(iterator.next());
+			first = HopRewriteUtils.createBinary(second, first, Hop.OpOp2.MULT);
+			first.setVisited();
+		}
+		return first;
+	}
+
+	private static Hop constructPower(Map.Entry<Hop, Integer> entry) {
+		final Hop hop = entry.getKey();
+		final int cnt = entry.getValue();
+		assert(cnt >= 1);
+		if (cnt == 1)
+			return hop; // don't set this visited... we will visit this next
+		Hop pow = HopRewriteUtils.createBinary(hop, new LiteralOp(cnt), Hop.OpOp2.POW);
+		pow.setVisited();
+		return pow;
+	}
+
+	/**
+	 * A Comparator that orders Hops by their data type, dimention, and sparsity.
+	 * The order is as follows:
+	 * 		scalars > row vectors > col vectors >
+	 *      non-vector matrices ordered by sparsity (higher nnz first, unknown sparsity last) >
+	 *      other data types.
+	 * Disambiguate by Hop ID.
+	 */
+	private static final Comparator<Hop> compareByDataType = new Comparator<Hop>() {
+		@Override
+		public final int compare(Hop o1, Hop o2) {
+			int c = Integer.compare(orderDataType[o1.getDataType().ordinal()], orderDataType[o2.getDataType().ordinal()]);
+			if (c != 0) return c;
+
+			// o1 and o2 have the same data type
+			switch (o1.getDataType()) {
+			case SCALAR: return Long.compare(o1.getHopID(), o2.getHopID());
+			case MATRIX:
+				// two matrices; check for vectors
+				if (o1.getDim1() == 1) { // row vector
+						if (o2.getDim1() != 1) return 1; // row vectors are greatest of matrices
+						return compareBySparsityThenId(o1, o2); // both row vectors
+				} else if (o2.getDim1() == 1) { // 2 is row vector; 1 is not
+						return -1; // row vectors are the greatest matrices
+				} else if (o1.getDim2() == 1) { // col vector
+						if (o2.getDim2() != 1) return 1; // col vectors greater than non-vectors
+						return compareBySparsityThenId(o1, o2); // both col vectors
+				} else if (o2.getDim2() == 1) { // 2 is col vector; 1 is not
+						return 1; // col vectors greater than non-vectors
+				} else { // both non-vectors
+						return compareBySparsityThenId(o1, o2);
+				}
+			default:
+				return Long.compare(o1.getHopID(), o2.getHopID());
+			}
+		}
+		private int compareBySparsityThenId(Hop o1, Hop o2) {
+			// the hop with more nnz is first; unknown nnz (-1) last
+			int c = Long.compare(o1.getNnz(), o2.getNnz());
+			if (c != 0) return c;
+			return Long.compare(o1.getHopID(), o2.getHopID());
+		}
+		private final int[] orderDataType;
+		{
+			Expression.DataType[] dtValues = Expression.DataType.values();
+			orderDataType = new int[dtValues.length];
+			for (int i = 0, valuesLength = dtValues.length; i < valuesLength; i++) {
```

Comment:
Yes, I do have a small lookup table; see `orderDataType`. The lookup table is not sufficient because we need to distinguish between two Hops that are both matrices.  Two matrices should be checked to see if they are (row or column) vectors, and further to see if we have sparsity information on them, since it is useful to group vectors together as well as to lead the emult with the sparsest matrices.